### PR TITLE
refactor(kb): unify ingest rollback facade

### LINKED ISSUE #570: ref(kb): C12.5 add operation rollback compatibility facade
# ref(kb): C12.5 add operation rollback compatibility facade

Part of `#494`.

Phase: Compatibility Layer

Depends on: `#497`, `#498`, `#499`, `#502`, `#503`, `#504`, `#506`

## Why

`#517` exposed that failed KB ingest can leave side effects across collection config, uploaded files, durable objects, document rows, parse/chunk rows, embeddings, status rows, and version state. The short-term fix can close the immediate holes, but the long-term rollback behavior needs one semantic operation boundary before the handle-replacement phase starts.

This issue adds a compatibility-layer rollback orchestration model without changing public API, tool, or pipeline result schemas. It should make rollback state explicit instead of inferring cleanup safety from counters such as `documents_created == 0`.

## Scope

Add coordinator-level operation orchestration for KB mutations:

- `KBOperationOutcome`
- `RollbackStatus`
- `PersistencePolicy`
- `SideEffectPlane`
- `CompensationStep`
- child operation aggregation for batch, cloud, and web ingest
- LIFO compensation execution
- idempotent compensation retries
- `side_effects_may_remain` reporting
- compatibility mapping back to existing API, tool, and pipeline result shapes

Route compatibility-layer ingest rollback through this operation model for:

- direct file ingest
- cloud ingest
- web ingest, with one child operation per URL/page
- background ingest handlers
- tool-created file and URL KBs

## Acceptance Criteria

- Rollback status is explicit: `not_needed`, `complete`, `incomplete`, or `skipped_by_policy`.
- Failed rollback cannot be treated as a clean failure.
- Metadata/config cleanup is blocked whenever `side_effects_may_remain` is true.
- Compensation steps are idempotent and can be retried safely.
- Batch operations can preserve successful child operations or roll them back according to an explicit `PersistencePolicy`.
- Direct, cloud, web, background, and tool ingest paths use the same operation outcome semantics.
- Existing public API response schemas, tool result models, pipeline return shapes, sync/async shapes, and exception mapping remain compatible.
- No FastAPI request/response objects are passed into `KBCoordinator`.

## Current Touchpoints

- coordinator and facade modules introduced by `#495` through `#506`
- `src/xagent/web/api/kb.py`
- `src/xagent/web/services/kb_file_service.py`
- `src/xagent/core/tools/core/RAG_tools/pipelines/document_ingestion.py`
- `src/xagent/core/tools/core/RAG_tools/pipelines/web_ingestion.py`
- `src/xagent/core/tools/core/RAG_tools/vector_storage/vector_manager.py`
- `src/xagent/core/tools/core/RAG_tools/version_management/`
- `src/xagent/core/tools/adapters/vibe/knowledge_tools.py`
- KB API, pipeline, tool, storage, and version-management tests

## Implementation Notes

- Treat rollback as a compensation/saga model, not as a distributed ACID transaction across SQL, LanceDB, durable storage, and filesystem state.
- Register compensation before or at the same boundary as the side effect it compensates.
- Execute compensation in reverse order.
- Do not delete collection config or metadata merely because no documents were successfully committed; use the rollback outcome.
- Use `#497` for uploaded-file, durable-object, and physical-file compensation.
- Use `#498`, `#499`, `#502`, and `#503` for compatibility-layer document/status/config/vector/version compensation until handle methods exist.
- Keep public compatibility wrappers as the external surface; new code should prefer the semantic coordinator operation API.

## Compatibility Invariants

- Current API-visible response bodies and HTTP status mapping remain route-layer concerns.
- Tool names, args schemas, and result model dumps do not change.
- Sync functions remain sync and async functions remain async.
- Existing retry/partial-success behavior is preserved unless a product decision explicitly changes it.
- Rollback incomplete state must be observable internally and must not be hidden by cleanup that removes the only evidence of remaining side effects.

## Validation

- Add unit tests for compensation registration order and reverse-order execution.
- Add unit tests for rollback failure reporting and idempotent retry.
- Add tests proving metadata/config cleanup is skipped when rollback is incomplete.
- Add direct ingest failure tests across config save, document registration, parse, chunk, and embedding boundaries.
- Add web ingest tests for one-page failure, multi-page partial success, zero-success rollback failure, and existing-file refresh rollback.
- Add cloud ingest mixed success/failure tests.
- Add background ingest parity tests proving sync and background paths use the same rollback policy.
- Add tool-created KB failure tests for file and URL ingestion.

## Handoff To Next Step

- `#507` should classify the new rollback operation types, wrappers, and any retained low-level compensation helpers.
- `#508` through `#513` should replace compatibility-layer compensation with handle-level document, parse/chunk, embedding, collection, status, and version compensation methods.
- `#514` should verify that coordinator rollback orchestration calls handles and file facades instead of backend-specific stores directly.
- `#515` should remove duplicated rollback policy from compatibility adapters after semantic ownership is complete.

## Non-Goals

- Do not introduce a production external vector backend.
- Do not change public API schemas or tool result schemas.
- Do not move parser, chunker, crawler, embedding provider, reranker, prompt formatting, or FastAPI concerns into the rollback operation model.
- Do not make physical lock/trash helpers collection-handle methods.

### PR DESCRIPTION
**Pull request by rogercloud**
## Summary
- Add API-level failed-ingest rollback result handling and cleanup decisions backed by `KBOperationOutcome`.
- Route direct, cloud, background, web, and tool-created ingest paths through the same rollback outcome semantics.
- Add LIFO/idempotent compensation execution and web page child-operation aggregation for rollback reporting.

Closes #570

## Tests
- `pre-commit run --all-files`
- `PYTHONPATH=src PYTHONDONTWRITEBYTECODE=1 /home/jicheng/github/xagent/.venv/bin/pytest -q tests/core/tools/core/RAG_tools/kb/test_api_compatibility.py tests/core/tools/core/RAG_tools/kb/test_operation_compatibility.py tests/core/tools/core/RAG_tools/kb/test_pipeline_compatibility.py tests/web/test_kb_ingest_lifecycle.py tests/core/tools/adapters/vibe/test_kb_creation_tools.py`
- `PYTHONPATH=src PYTHONDONTWRITEBYTECODE=1 /home/jicheng/github/xagent/.venv/bin/pytest -q tests/web/test_background_jobs.py -k 'not celery_worker_app_import_registers_tasks and not trigger_event_job_runs_with_eager_celery'`

### CODE REVIEW

File: src/xagent/core/tools/core/RAG_tools/kb/pipeline_compatibility.py [hunk 1]
```diff
@@ -403,11 +423,16 @@ def finish_web_page_operation(
...
-        rollback_status = (
-            RollbackStatus.NOT_NEEDED
-            if status == "success" or not side_effects_may_remain
-            else RollbackStatus.INCOMPLETE
-        )
+        if status == "success":
+            rollback_status = RollbackStatus.NOT_NEEDED
+        elif side_effects_may_remain:
+            rollback_status = RollbackStatus.INCOMPLETE
+        elif operation.compensation_attempted and operation.has_side_effects():
```
Comment (qinxuye) [inline src/xagent/core/tools/core/RAG_tools/kb/pipeline_compatibility.py]: [P1] Don’t mark the whole page rollback complete just because one file compensation ran. A web page operation also records document/status/parse/vector side effects from the nested document ingestion, but those steps do not have completed compensation callbacks here. With a custom `rollback_on_failure` that only cleans file persistence, or the current test double that just records the callback invocation, this branch reports `RollbackStatus.COMPLETE` and `side_effects_may_remain=False` while the document side effects are still present in the outcome. Please either model the web-page rollback callback as compensating those document planes too, or only report `COMPLETE` once all recorded side-effect planes have been covered by successful compensation.

**Review by qinxuye — COMMENTED** [review]
I found one rollback-state issue that should be addressed before relying on the new facade semantics.

File: src/xagent/core/tools/core/RAG_tools/kb/pipeline_compatibility.py [hunk 2]
```diff
@@ -331,7 +333,17 @@ async def run_web_ingestion(
                     pipeline_facade=self,
                 )
                 await self.ensure_collection_backend_binding_async(collection)
-                self._record_web_ingestion_outcome(operation, result)
+                outcome = self._record_web_ingestion_outcome(operation, result)
+                if (
```
Comment (qinxuye) [inline src/xagent/core/tools/core/RAG_tools/kb/pipeline_compatibility.py:338]: [P1] This now propagates an overly conservative page outcome back to the public result. The production web `rollback_on_failure` callbacks clean both the persisted web file and the RAG/document writes, but the operation model only marks the FILE compensation step as completed; the document/status/parse/vector steps recorded from the nested document ingestion remain `uncompensated_steps()`. As a result, a fully successful web-page rollback still makes `outcome.side_effects_may_remain=True`, this assignment copies that to `WebIngestionResult`, and the API/background failed-ingest cleanup path will skip new-collection metadata/config cleanup even though the rollback actually succeeded. Please either register/complete compensation coverage for the document-side steps when the web rollback callback performs that cleanup, or avoid copying this conservative internal outcome into the cleanup decision until the coverage model can distinguish truly remaining side effects.

**Review by qinxuye — COMMENTED** [review]
I found one follow-up issue in the updated rollback coverage semantics.

**Review by qinxuye — APPROVED** [review]
LGTM
